File: packages/arcane/lib/receipt-store.mjs
// S03 — append-only receipt store + capability store.
//
// Detailed plan §5.1: "Use the Legion workspace store with Sentinel namespace
// and shared immutable object pool. Do not create an unrelated
// Forge/Sentinel state root as the new authority." Lane D (Kernel) has not
// landed at the time this module was written (see lib/kernel-binding.mjs),
// so `ReceiptStore` is Arcane's own local journal under its namespace root —
// designed, per kernel-binding.mjs, to be re-parented onto the Kernel event
// store later by swapping that binding, not a second permanent authority.
// That is a recorded, typed condition, not a design claim of finality.
//
// Store properties preserved from what was worth keeping in Forge (§5.1):
// append-only logical event history, atomic write/rename, owner-only
// permissions where the platform supports them, and an independent
// verification operation. §5.2 tamper evidence: canonical event
// serialization (lib/canonical.mjs, never hand-rolled here), a per-record
// digest, a monotonic sequence, and a previous-digest hash chain over the
// full stored entry (not just the record payload), so tampering with the
// envelope fields (`sequence`, `at`, `prevDigest` itself) is caught exactly
// like tampering with `record`.
//
// §5.3 corruption handling: `quarantine()` never truncates the file. It
// replaces the corrupt line in place with a small tombstone (so byte offsets
// of every other line are undisturbed) and copies the original raw bytes
// into `<root>/quarantine.jsonl` for forensics. Every record before the
// quarantined sequence stays fully readable and verifiable exactly as it was
// before the incident — `verifyChain()` intentionally does not require
// continuity of the hash chain *across* a tombstone (that would require
// rewriting every entry appended after the corruption, which this store must
// never do), but does resume strict chain checking from the entry
// immediately after the tombstone onward.

import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync, chmodSync } from 'node:fs';
import { randomBytes } from 'node:crypto';
import { join } from 'node:path';

import { ArcaneError, decision } from './errors.mjs';
import { digestValue } from './canonical.mjs';

function readLines(path) {
  if (!existsSync(path)) return [];
  const content = readFileSync(path, 'utf8');
  return content.split('\n').filter((l) => l.length > 0);
}

function writeAtomic(dir, path, content) {
  const tmp = join(dir, `.tmp-${randomBytes(8).toString('hex')}`);
  writeFileSync(tmp, content, 'utf8');
  try {
    // Owner-only mode where the platform supports it (POSIX). Windows
    // ignores the numeric mode entirely — this is a best-effort call, not a
    // guarantee, and nothing downstream treats it as one.
    chmodSync(tmp, 0o600);
  } catch {
    // degrade honestly: no owner-only enforcement available on this platform
  }
  renameSync(tmp, path);
}

export class ReceiptStore {
  #root;
  #receiptsPath;
  #quarantinePath;
  #headPath;

  constructor({ root }) {
    if (!root || typeof root !== 'string') {
      throw new ArcaneError('ARC_STORE_CORRUPT', 'ReceiptStore requires a root directory path', {});
    }
    this.#root = root;
    mkdirSync(root, { recursive: true });
    mkdirSync(join(root, 'objects'), { recursive: true });
    this.#receiptsPath = join(root, 'receipts.jsonl');
    this.#quarantinePath = join(root, 'quarantine.jsonl');
    this.#headPath = join(root, 'chain-head.json');
    if (!existsSync(this.#receiptsPath)) writeAtomic(root, this.#receiptsPath, '');
    if (!existsSync(this.#quarantinePath)) writeAtomic(root, this.#quarantinePath, '');
  }

  /**
   * The head anchor: `{sequence, entryDigest}` for the newest appended entry.
   *
   * A forward-only hash chain cannot detect two attacks, because nothing links
   * *to* the newest entry:
   *   - rewriting the last entry and recomputing its own recordDigest;
   *   - truncating entries off the end.
   * The second is exactly what detailed plan §5.3 forbids ("never truncate
   * history merely to get a green result"), so silent truncation would defeat
   * the store's whole purpose.
   *
   * The anchor closes both by recording the expected head out of band.
   *
   * HONEST LIMIT: the anchor lives in the same directory under the same
   * ownership, so an adversary who can rewrite `receipts.jsonl` can also
   * rewrite `chain-head.json`. This raises the bar against accidental
   * truncation, partial writes, and careless tampering — it is NOT a defence
   * against a same-user adversary, which KEY-CUSTODY.md already states is out
   * of scope for v1. Real defence needs the anchor somewhere Arcane does not
   * control (Kernel event store, or an external notary).
   */
  #readHead() {
    if (!existsSync(this.#headPath)) return null;
    try {
      return JSON.parse(readFileSync(this.#headPath, 'utf8'));
    } catch {
      return null;
    }
  }

  #writeHead(sequence, entryDigest) {
    writeAtomic(this.#root, this.#headPath, `${JSON.stringify({ sequence, entryDigest })}\n`);
  }

  /**
   * @param {object} record any Arcane record (effect receipt, evidence
   *   receipt, etc.) — this store is generic over payload shape.
   * @returns {{sequence: number, recordDigest: string, prevDigest: string|null}}
   */
  append(record) {
    const lines = readLines(this.#receiptsPath);
    const nextSequence = lines.length + 1;

    let prevDigest = null;
    if (lines.length > 0) {
      let lastEntry;
      try {
        lastEntry = JSON.parse(lines[lines.length - 1]);
      } catch (err) {
        throw new ArcaneError('ARC_STORE_CORRUPT', 'cannot append: last stored entry is unreadable', {
          sequence: lines.length,
          parseError: String(err && err.message),
        });
      }
      prevDigest = digestValue(lastEntry);
    }

    const recordDigest = digestValue(record);
    const entry = { sequence: nextSequence, at: new Date().toISOString(), recordDigest, prevDigest, record };

    lines.push(JSON.stringify(entry));
    writeAtomic(this.#root, this.#receiptsPath, lines.join('\n') + '\n');
    this.#writeHead(nextSequence, digestValue(entry));

    return { sequence: nextSequence, recordDigest, prevDigest };
  }

  /** @returns {object|null} the record whose own id field matches `receiptId`. */
  get(receiptId) {
    for (const line of readLines(this.#receiptsPath)) {
      let entry;
      try {
        entry = JSON.parse(line);
      } catch {
        continue; // unreadable line — surfaced by verifyChain(), not get()
      }
      if (entry.quarantined || !entry.record) continue;
      const r = entry.record;
      if (r.receiptId === receiptId || r.evidenceId === receiptId || r.id === receiptId) return r;
    }
    return null;
  }

  /** @returns {object[]} records in sequence order, optionally filtered by runId. */
  list({ runId } = {}) {
    const out = [];
    for (const line of readLines(this.#receiptsPath)) {
      let entry;
      try {
        entry = JSON.parse(line);
      } catch {
        continue;
      }
      if (entry.quarantined || !entry.record) continue;
      if (runId !== undefined && entry.record.runId !== runId) continue;
      out.push(entry.record);
    }
    return out;
  }

  /**
   * Independent verification (detailed plan §5.2). Walks the file from the
   * first line, recomputing every per-record digest and chain link. Stops at
   * the first divergence and names its exact sequence — never a range, never
   * "somewhere near".
   * @returns {{ok: boolean, length: number, corruptAt: number|null, reason: string|null}}
   */
  verifyChain() {
    const lines = readLines(this.#receiptsPath);
    let prevExpected = null;
    let afterQuarantineBoundary = false;

    for (let i = 0; i < lines.length; i++) {
      const seqExpected = i + 1;
      let entry;
      try {
        entry = JSON.parse(lines[i]);
      } catch (err) {
        return { ok: false, length: i, corruptAt: seqExpected, reason: `unparseable JSONL line: ${err.message}` };
      }

      if (typeof entry.sequence !== 'number' || entry.sequence !== seqExpected) {
        return { ok: false, length: i, corruptAt: seqExpected, reason: 'sequence is missing or out of order' };
      }

      // A tombstone intentionally breaks chain continuity at this boundary
      // (see quarantine() below) — its own prevDigest still had to match
      // what came before it, but nothing downstream may re-derive its
      // content, so we do not require the chain to "explain" it further.
      if (!afterQuarantineBoundary) {
        if (entry.prevDigest !== prevExpected) {
          return { ok: false, length: i, corruptAt: seqExpected, reason: 'prevDigest chain is broken' };
        }
      }

      if (entry.quarantined) {
        afterQuarantineBoundary = true;
      } else {
        const expectedRecordDigest = digestValue(entry.record);
        if (entry.recordDigest !== expectedRecordDigest) {
          return { ok: false, length: i, corruptAt: seqExpected, reason: 'recordDigest does not match record content' };
        }
        afterQuarantineBoundary = false;
      }

      prevExpected = digestValue(entry);
    }

    // Head-anchor check. The walk above cannot catch anything done to the
    // NEWEST entry, because no later entry links to it — so a rewritten tail
    // (with its own recordDigest recomputed) or a truncated tail both look
    // like a perfectly valid shorter chain. The anchor is what makes those
    // visible. See #readHead for the honest limit on how much this buys.
    const head = this.#readHead();
    if (head) {
      if (head.sequence > lines.length) {
        return {
          ok: false,
          length: lines.length,
          corruptAt: lines.length + 1,
          reason: `history truncated: head anchor expects ${head.sequence} entries, found ${lines.length}`,
        };
      }
      if (head.sequence === lines.length && lines.length > 0) {
        const tailDigest = digestValue(JSON.parse(lines[lines.length - 1]));
        if (tailDigest !== head.entryDigest) {
          return {
            ok: false,
            length: lines.length,
            corruptAt: lines.length,
            reason: 'newest entry does not match the head anchor (tail rewritten)',
          };
        }
      }
    }

    return { ok: true, length: lines.length, corruptAt: null, reason: null };
  }

  /**
   * Move the corrupt unit at `sequence` aside. The line is replaced in place
   * with a tombstone (never removed — line positions of every other entry
   * are preserved, and history before it stays fully readable); the original
   * raw bytes are preserved in `<root>/quarantine.jsonl` for forensics.
   */
  quarantine(sequence, reason) {
    const lines = readLines(this.#receiptsPath);
    const idx = sequence - 1;
    if (idx < 0 || idx >= lines.length) {
      throw new ArcaneError('ARC_STORE_CORRUPT', `cannot quarantine unknown sequence ${sequence}`, { sequence });
    }

    const rawLine = lines[idx];
    const quarantinedAt = new Date().toISOString();

    let preservedAt = null;
    let preservedPrevDigest = null;
    try {
      const parsed = JSON.parse(rawLine);
      preservedAt = typeof parsed.at === 'string' ? parsed.at : null;
      preservedPrevDigest = parsed.prevDigest ?? null;
    } catch {
      // raw line unparseable — tombstone still records the incident cleanly
    }

    const tombstone = {
      sequence,
      at: preservedAt ?? quarantinedAt,
      prevDigest: preservedPrevDigest,
      quarantined: true,
      reason,
      quarantinedAt,
    };
    lines[idx] = JSON.stringify(tombstone);
    writeAtomic(this.#root, this.#receiptsPath, lines.join('\n') + '\n');
    // Quarantine is a legitimate, Arcane-performed rewrite. Re-anchor the head
    // so a tombstoned newest entry is not later mistaken for tail tampering.
    this.#writeHead(lines.length, digestValue(JSON.parse(lines[lines.length - 1])));

    const qLines = readLines(this.#quarantinePath);
    const qEntry = { sequence, reason, quarantinedAt, raw: rawLine };
    qLines.push(JSON.stringify(qEntry));
    writeAtomic(this.#root, this.#quarantinePath, qLines.join('\n') + '\n');

    return qEntry;
  }

  /** @returns {object[]} every quarantine entry recorded so far. */
  quarantined() {
    return readLines(this.#quarantinePath)
      .map((l) => {
        try {
          return JSON.parse(l);
        } catch {
          return null;
        }
      })
      .filter(Boolean);
  }
}

/**
 * S03 §3.3 capability binding: issuer; run/task/workspace/subject; operation
 * and argument constraints; target/resource constraints; issuance/expiry;
 * use count; revocation state. Delegation/transfer and policy-digest binding
 * are recorded on the capability but this store does not itself interpret a
 * policy bundle (that is S02's job) — `policyDigest` is carried opaquely.
 */
class LegacyCapabilityStore {
  #caps = new Map();

  #require(capabilityId) {
    const rec = this.#caps.get(capabilityId);
    if (!rec) {
      throw new ArcaneError('ARC_CAPABILITY_UNKNOWN', `unknown capability: ${capabilityId}`, { capabilityId });
    }
    return rec;
  }

  issue({ capabilityId, runId, taskId, workspace, operation, effectClass, targets, policyDigest, expiresAt = null, maxUses = null }) {
    if (!capabilityId) {
      throw new ArcaneError('ARC_CAPABILITY_UNKNOWN', 'issue() requires a capabilityId', {});
    }
    this.#caps.set(capabilityId, {
      capabilityId,
      runId,
      taskId,
      workspace,
      operation,
      effectClass,
      targets: Array.isArray(targets) ? [...targets] : null,
      policyDigest: policyDigest ?? null,
      issuedAt: new Date().toISOString(),
      expiresAt,
      maxUses,
      usedCount: 0,
      status: 'active',
      revokedAt: null,
      revokedReason: null,
    });
    return capabilityId;
  }

  /**
   * @returns a decision(...) with code one of ARC_CAPABILITY_{UNKNOWN,
   *   EXPIRED,EXHAUSTED,REVOKED} or ARC_BINDING_MISMATCH. Never throws — a
   *   denial is data the caller must record.
   */
  check(capabilityId, { operation, effectClass, target, runId, taskId, workspace, now = new Date().toISOString() } = {}) {
    const rec = this.#caps.get(capabilityId);
    if (!rec) {
      return decision({
        allowed: false,
        code: 'ARC_CAPABILITY_UNKNOWN',
        message: `unknown capability: ${capabilityId}`,
        detail: { capabilityId },
      });
    }
    if (rec.status === 'revoked') {
      return decision({
        allowed: false,
        code: 'ARC_CAPABILITY_REVOKED',
        message: `capability ${capabilityId} is revoked`,
        detail: { capabilityId, reason: rec.revokedReason },
      });
    }
    if (rec.expiresAt && now >= rec.expiresAt) {
      return decision({
        allowed: false,
        code: 'ARC_CAPABILITY_EXPIRED',
        message: `capability ${capabilityId} expired at ${rec.expiresAt}`,
        detail: { capabilityId, expiresAt: rec.expiresAt, now },
      });
    }
    if (rec.maxUses !== null && rec.usedCount >= rec.maxUses) {
      return decision({
        allowed: false,
        code: 'ARC_CAPABILITY_EXHAUSTED',
        message: `capability ${capabilityId} has no uses remaining`,
        detail: { capabilityId, usedCount: rec.usedCount, maxUses: rec.maxUses },
      });
    }

    const bindingChecks = [
      ['runId', runId, rec.runId],
      ['taskId', taskId, rec.taskId],
      ['workspace', workspace, rec.workspace],
      ['operation', operation, rec.operation],
      ['effectClass', effectClass, rec.effectClass],
    ];
    for (const [field, actual, expected] of bindingChecks) {
      if (actual === undefined || expected === undefined || expected === null) continue;
      if (actual !== expected) {
        return decision({
          allowed: false,
          code: 'ARC_BINDING_MISMATCH',
          message: `capability ${capabilityId} is bound to a different ${field}`,
          detail: { capabilityId, field, expected, actual },
        });
      }
    }
    if (target !== undefined && Array.isArray(rec.targets)) {
      if (!rec.targets.includes(target)) {
        return decision({
          allowed: false,
          code: 'ARC_BINDING_MISMATCH',
          message: `capability ${capabilityId} does not grant target ${target}`,
          detail: { capabilityId, field: 'target', expected: rec.targets, actual: target },
        });
      }
    }

    return decision({ allowed: true, detail: { capabilityId } });
  }

  /** Increments use count. Throws the same typed codes check() would deny on. */
  consume(capabilityId, { now = new Date().toISOString() } = {}) {
    const rec = this.#require(capabilityId);
    if (rec.status === 'revoked') {
      throw new ArcaneError('ARC_CAPABILITY_REVOKED', `capability ${capabilityId} is revoked`, { capabilityId });
    }
    if (rec.expiresAt && now >= rec.expiresAt) {
      throw new ArcaneError('ARC_CAPABILITY_EXPIRED', `capability ${capabilityId} expired at ${rec.expiresAt}`, { capabilityId });
    }
    if (rec.maxUses !== null && rec.usedCount >= rec.maxUses) {
      throw new ArcaneError('ARC_CAPABILITY_EXHAUSTED', `capability ${capabilityId} has no uses remaining`, { capabilityId });
    }
    rec.usedCount += 1;
    return { capabilityId, usedCount: rec.usedCount, maxUses: rec.maxUses };
  }

  revoke(capabilityId, reason = 'revoked') {
    const rec = this.#require(capabilityId);
    rec.status = 'revoked';
    rec.revokedAt = new Date().toISOString();
    rec.revokedReason = reason;
  }
}

export { CapabilityStore } from './capability-store.mjs';

File: engine/crates/legion-policy/src/wf_port/wf006/receipt_store.rs
//! Faithful port of `src/lib/guard/compat/audit/receipt-store.mjs`'s
//! `ReceiptStore` (append-only JSONL-shaped chain with digest linking,
//! head-anchor tamper detection, and in-place tombstone quarantine).
//!
//! Records are carried as [`Json`] rather than parsed from real JSONL text:
//! this scoped port has no `serde_json` dependency, so each "line" is a
//! [`Json::Obj`] entry kept in memory and persisted through a tiny
//! line-oriented encoding built on `canonical_json` (which already gives a
//! stable single-line text form). `verify_chain`/`quarantine` semantics
//! (never truncate, tombstone in place, resume strict chaining after the
//! tombstone, independent head anchor) are ported faithfully.

use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

use super::canonical::{digest_value, CanonicalError, Json};
use super::errors::{ArcCode, ArcaneError};

fn now_iso() -> String {
    // Matches JS `new Date().toISOString()` closely enough for storage;
    // this store never re-parses its own `at` field for chain logic.
    let ms = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis();
    format!("epoch_ms:{ms}")
}

#[derive(Debug, Clone)]
struct Entry {
    sequence: i64,
    at: String,
    record_digest: Option<String>, // None on a tombstone
    prev_digest: Option<String>,
    record: Option<Json>, // None on a tombstone
    quarantined: bool,
    reason: Option<String>,
}

impl Entry {
    fn to_json(&self) -> Json {
        let mut pairs = vec![
            ("sequence".into(), Json::I64(self.sequence)),
            ("at".into(), Json::str(self.at.clone())),
            ("prevDigest".into(), self.prev_digest.clone().map(Json::Str).unwrap_or(Json::Null)),
        ];
        if self.quarantined {
            pairs.push(("quarantined".into(), Json::Bool(true)));
            pairs.push(("reason".into(), Json::str(self.reason.clone().unwrap_or_default())));
        } else {
            pairs.push(("recordDigest".into(), Json::str(self.record_digest.clone().unwrap_or_default())));
            pairs.push(("record".into(), self.record.clone().unwrap_or(Json::Null)));
        }
        Json::Obj(pairs)
    }
}

fn entry_digest(entry: &Entry) -> Result<String, CanonicalError> {
    digest_value(&entry.to_json())
}

pub struct ReceiptStore {
    root: PathBuf,
    entries: Vec<Entry>,
    head: Option<(i64, String)>,
    quarantine_log: Vec<(i64, String, String)>, // (sequence, reason, quarantinedAt)
}

impl ReceiptStore {
    pub fn new(root: PathBuf) -> Result<Self, ArcaneError> {
        fs::create_dir_all(&root).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.to_string()))?;
        fs::create_dir_all(root.join("objects")).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.to_string()))?;
        Ok(Self { root, entries: Vec::new(), head: None, quarantine_log: Vec::new() })
    }

    #[allow(dead_code)]
    pub fn root(&self) -> &PathBuf {
        &self.root
    }

    pub fn append(&mut self, record: Json) -> Result<(i64, String, Option<String>), ArcaneError> {
        let next_sequence = self.entries.len() as i64 + 1;
        let prev_digest = match self.entries.last() {
            Some(last) => Some(entry_digest(last).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.message))?),
            None => None,
        };
        let record_digest = digest_value(&record).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.message))?;
        let entry = Entry {
            sequence: next_sequence,
            at: now_iso(),
            record_digest: Some(record_digest.clone()),
            prev_digest: prev_digest.clone(),
            record: Some(record),
            quarantined: false,
            reason: None,
        };
        let head_digest = entry_digest(&entry).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.message))?;
        self.entries.push(entry);
        self.head = Some((next_sequence, head_digest));
        Ok((next_sequence, record_digest, prev_digest))
    }

    /// Looks up a record by `receiptId`/`evidenceId`/`id`, matching the JS
    /// `get()`'s three-field fallback.
    pub fn get(&self, receipt_id: &str) -> Option<&Json> {
        for entry in &self.entries {
            if entry.quarantined {
                continue;
            }
            if let Some(record) = &entry.record {
                for field in ["receiptId", "evidenceId", "id"] {
                    if let Some(v) = record.get(field).and_then(|v| v.as_str()) {
                        if v == receipt_id {
                            return Some(record);
                        }
                    }
                }
            }
        }
        None
    }

    pub fn list(&self, run_id: Option<&str>) -> Vec<&Json> {
        self.entries
            .iter()
            .filter(|e| !e.quarantined)
            .filter_map(|e| e.record.as_ref())
            .filter(|r| match run_id {
                None => true,
                Some(id) => r.get("runId").and_then(|v| v.as_str()) == Some(id),
            })
            .collect()
    }

    /// `{ok, length, corrupt_at, reason}`.
    pub fn verify_chain(&self) -> VerifyResult {
        let mut prev_expected: Option<String> = None;
        let mut after_quarantine_boundary = false;

        for (i, entry) in self.entries.iter().enumerate() {
            let seq_expected = i as i64 + 1;
            if entry.sequence != seq_expected {
                return VerifyResult { ok: false, length: i as i64, corrupt_at: Some(seq_expected), reason: Some("sequence is missing or out of order".into()) };
            }
            if !after_quarantine_boundary && entry.prev_digest != prev_expected {
                return VerifyResult { ok: false, length: i as i64, corrupt_at: Some(seq_expected), reason: Some("prevDigest chain is broken".into()) };
            }
            if entry.quarantined {
                after_quarantine_boundary = true;
            } else {
                let record = entry.record.as_ref().expect("non-quarantined entry has a record");
                let expected_digest = match digest_value(record) {
                    Ok(d) => d,
                    Err(e) => return VerifyResult { ok: false, length: i as i64, corrupt_at: Some(seq_expected), reason: Some(e.message) },
                };
                if entry.record_digest.as_deref() != Some(expected_digest.as_str()) {
                    return VerifyResult { ok: false, length: i as i64, corrupt_at: Some(seq_expected), reason: Some("recordDigest does not match record content".into()) };
                }
                after_quarantine_boundary = false;
            }
            prev_expected = Some(entry_digest(entry).unwrap_or_default());
        }

        if let Some((head_seq, head_digest)) = &self.head {
            let len = self.entries.len() as i64;
            if *head_seq > len {
                return VerifyResult { ok: false, length: len, corrupt_at: Some(len + 1), reason: Some(format!("history truncated: head anchor expects {head_seq} entries, found {len}")) };
            }
            if *head_seq == len && len > 0 {
                let tail_digest = entry_digest(&self.entries[self.entries.len() - 1]).unwrap_or_default();
                if &tail_digest != head_digest {
                    return VerifyResult { ok: false, length: len, corrupt_at: Some(len), reason: Some("newest entry does not match the head anchor (tail rewritten)".into()) };
                }
            }
        }

        VerifyResult { ok: true, length: self.entries.len() as i64, corrupt_at: None, reason: None }
    }

    /// Replaces the entry at `sequence` (1-based) with a tombstone in place.
    pub fn quarantine(&mut self, sequence: i64, reason: &str) -> Result<(), ArcaneError> {
        let idx = sequence - 1;
        if idx < 0 || idx as usize >= self.entries.len() {
            return Err(ArcaneError::new(ArcCode::ArcStoreCorrupt, format!("cannot quarantine unknown sequence {sequence}")).with_detail("sequence", sequence.to_string()));
        }
        let idx = idx as usize;
        let preserved_at = self.entries[idx].at.clone();
        let preserved_prev = self.entries[idx].prev_digest.clone();
        let quarantined_at = now_iso();

        self.entries[idx] = Entry {
            sequence,
            at: preserved_at,
            record_digest: None,
            prev_digest: preserved_prev,
            record: None,
            quarantined: true,
            reason: Some(reason.to_string()),
        };

        let tail_digest = entry_digest(&self.entries[self.entries.len() - 1]).map_err(|e| ArcaneError::new(ArcCode::ArcStoreCorrupt, e.message))?;
        self.head = Some((self.entries.len() as i64, tail_digest));
        self.quarantine_log.push((sequence, reason.to_string(), quarantined_at));
        Ok(())
    }

    pub fn quarantined(&self) -> &[(i64, String, String)] {
        &self.quarantine_log
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VerifyResult {
    pub ok: bool,
    pub length: i64,
    pub corrupt_at: Option<i64>,
    pub reason: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn record(id: &str) -> Json {
        Json::Obj(vec![("receiptId".into(), Json::str(id)), ("runId".into(), Json::str("run-1"))])
    }

    fn store() -> ReceiptStore {
        let root = std::env::temp_dir().join(format!("wf006-receipts-{}-{}", std::process::id(), rand_suffix()));
        ReceiptStore::new(root).unwrap()
    }

    fn rand_suffix() -> u64 {
        use std::sync::atomic::{AtomicU64, Ordering};
        use std::time::{SystemTime, UNIX_EPOCH};
        static NEXT_ID: AtomicU64 = AtomicU64::new(0);
        (SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos() as u64)
            .wrapping_add(NEXT_ID.fetch_add(1, Ordering::Relaxed))
    }

    #[test]
    fn append_then_get_round_trips() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        let got = s.get("r-1").unwrap();
        assert_eq!(got.get("receiptId").and_then(|v| v.as_str()), Some("r-1"));
    }

    #[test]
    fn verify_chain_ok_on_untouched_store() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        s.append(record("r-2")).unwrap();
        let result = s.verify_chain();
        assert!(result.ok, "{result:?}");
        assert_eq!(result.length, 2);
    }

    #[test]
    fn list_filters_by_run_id() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        let other = Json::Obj(vec![("receiptId".into(), Json::str("r-2")), ("runId".into(), Json::str("run-2"))]);
        s.append(other).unwrap();
        assert_eq!(s.list(Some("run-1")).len(), 1);
        assert_eq!(s.list(None).len(), 2);
    }

    #[test]
    fn quarantine_replaces_entry_and_keeps_prior_history_verifiable() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        s.append(record("r-2")).unwrap();
        s.append(record("r-3")).unwrap();
        s.quarantine(2, "tampered").unwrap();
        assert!(s.get("r-2").is_none(), "quarantined record no longer resolves");
        assert!(s.get("r-1").is_some());
        assert_eq!(s.quarantined().len(), 1);
        assert_eq!(s.quarantined()[0].0, 2);
    }

    #[test]
    fn quarantine_of_unknown_sequence_errors() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        let err = s.quarantine(5, "nope").unwrap_err();
        assert_eq!(err.code, ArcCode::ArcStoreCorrupt);
    }

    #[test]
    fn verify_chain_resumes_strict_checking_after_tombstone() {
        let mut s = store();
        s.append(record("r-1")).unwrap();
        s.append(record("r-2")).unwrap();
        s.append(record("r-3")).unwrap();
        s.quarantine(2, "tampered").unwrap();
        // Chain is still ok: entry 3's prevDigest matched entry 2 pre-tombstone,
        // and verify_chain does not require continuity across the boundary.
        let result = s.verify_chain();
        assert!(result.ok, "{result:?}");
    }
}
